cpp/ql/lib/semmle/code/cpp/models/interfaces/Sql.qll

```diff
@@ -0,0 +1,30 @@
+/**
+ * Provides abstract classes for modeling functions that execute and escape SQL query strings.
+ * To use this QL library, create a QL class extending `SqlSink` or `SqlBarrier` with a
+ * characteristic predicate that selects the function or set of functions you are modeling.
+ * Within that class, override the predicates provided by the class to match the way a
+ * parameter flows into the function and, in the case of `SqlBarrier`, out of the function.
+ */
+
+private import cpp
+
+/**
+ * An abstract class that represents a function that executes an SQL query.
+ */
+abstract class SqlSink extends Function {
+  /**
+   * Holds if `input` to this function represents data that is passed to an SQL server.
```

Comment:
Should this be specifically an input that's interpreted as a SQL query string, as opposed to arguments to a prepared statement?

Reply:
Good question. I think a function producing a prepared statement should be a sanitizer, and the function that consumes the prepared statement should be an `SqlSink`. Do you think this distinction be made more clear in the QLDoc?

Reply:
Should "data that is passed to an SQL server" instead be "SQL code to be executed"? Then the distinction is clear to me: a prepared statement that's been compiled is not _SQL code_. I'm not sure if _SQL code_ is exactly the right term, but I want to make sure we include `UPDATE` etc. and not just `SELECT`.

Reply:
Good point. I like that description better. I'll update the QLDoc with that phrasing.

Reply:
Fixed in 947ab8a14d4357a69ed736a2ba9df2e82b09f339.

Reply:
To be clear, I think the string argument to a function that *produces* a prepared statement should be a SQL injection sink (if it's tainted, then an attacker has more control over the SQL that's executed than they were intended to), and nothing in the function that actually executes the prepared statement should be a sink (if the prepared statement is tainted, that's handled already, and the other arguments aren't parsed as SQL code).